File: src/main/java/frc/team4069/robot/Robot.java
package frc.team4069.robot;


import edu.wpi.first.wpilibj.IterativeRobot;
import edu.wpi.first.wpilibj.command.Scheduler;
import frc.team4069.robot.commands.CommandBase;
import frc.team4069.robot.commands.OperatorControlCommandGroup;
import frc.team4069.robot.commands.autonomous.DriveToSwitchCommand;
import frc.team4069.robot.io.Input;
import frc.team4069.robot.subsystems.ArmSubsystem;
import frc.team4069.robot.subsystems.ElevatorSubsystem;

public class Robot extends IterativeRobot {

    private Scheduler scheduler;

    @Override
    public void robotInit() {
        super.robotInit();

        // Set up the input class
        Input.init();

        // Configure the vision threads
//        VisionData.configureVision();

        // Get the scheduler
        scheduler = Scheduler.getInstance();
        // Initialize the subsystems
        CommandBase.init();
    }

    @Override
    public void autonomousInit() {
        super.autonomousInit();
        // Drive forward and turn towards the switch
        scheduler.add(new DriveToSwitchCommand());
    }

    @Override
    public void teleopInit() {
        super.teleopInit();
        // Remove all commands from the scheduler so no autonomous tasks continue running
        scheduler.removeAll();
        // Add an operator control command group to the scheduler, which should never exit
        scheduler.add(new OperatorControlCommandGroup());
    }

    @Override
    public void disabledInit() {
        // Reset the state of the elevator subsystem so that it doesn't take off when next we enable
        ElevatorSubsystem.INSTANCE.reset();
        ArmSubsystem.INSTANCE.reset();
    }

    // During all phases, run the command scheduler
    private void universalPeriodic() {
        scheduler.run();
    }

    @Override
    public void autonomousPeriodic() {
        universalPeriodic();
    }

    @Override
    public void teleopPeriodic() {
        super.teleopPeriodic();
        universalPeriodic();
    }
}

File: src/main/java/frc/team4069/robot/Robot.kt
package frc.team4069.robot

import edu.wpi.first.wpilibj.IterativeRobot
import edu.wpi.first.wpilibj.command.Scheduler
import frc.team4069.robot.commands.CommandBase
import frc.team4069.robot.commands.OperatorControlCommandGroup
import frc.team4069.robot.commands.arm.StartArmCommand
import frc.team4069.robot.io.Input
import frc.team4069.robot.subsystems.ArmSubsystem
import frc.team4069.robot.subsystems.ElevatorSubsystem
import frc.team4069.robot.subsystems.VacuumSubsystem
import frc.team4069.robot.subsystems.WinchSubsystem
import frc.team4069.robot.vision.VisionData

class Robot : IterativeRobot() {
    private lateinit var scheduler: Scheduler

    // Initialise the various portions of the robot ie vision, command scheduler, IO
    override fun robotInit() {
        Input.init()

        VisionData.configureVision()
        scheduler = Scheduler.getInstance()
        CommandBase.init()
    }

    // Schedule autonomous actions
    override fun autonomousInit() {
        //TODO: Autonomous operations
        scheduler.add(StartArmCommand())
    }

    // Schedule teleop command group
    override fun teleopInit() {
        scheduler.removeAll()
        scheduler.add(OperatorControlCommandGroup())
    }

    // Reset the encoders and motors of our subsystems
    override fun disabledInit() {
        ElevatorSubsystem.reset()
        ArmSubsystem.reset()
        WinchSubsystem.reset()
        VacuumSubsystem.reset()
    }

    // Periodic operations; just run scheduler
    override fun autonomousPeriodic() = scheduler.run()
    override fun teleopPeriodic() = scheduler.run()
}